crates/autopilot/src/infra/persistence/mod.rs

```diff
@@ -141,38 +145,198 @@ impl Persistence {
             .context("begin")
-            .map_err(Error::DbError)?;
+            .map_err(error::Auction::DbError)?;
 
         let db_prices = database::auction_prices::fetch(&mut ex, auction)
             .await
             .context("fetch")
-            .map_err(Error::DbError)?;
+            .map_err(error::Auction::DbError)?;
 
         let mut prices = HashMap::new();
         for price in db_prices {
             let token = eth::H160(price.token.0).into();
             let price = big_decimal_to_u256(&price.price)
                 .ok_or(domain::auction::InvalidPrice)
                 .and_then(|p| domain::auction::Price::new(p.into()))
-                .map_err(Error::Price)?;
+                .map_err(error::Auction::Price)?;
             prices.insert(token, price);
         }
 
         Ok(prices)
     }
+
+    /// Get auction data related to the given settlement.
+    pub async fn get_settlement_auction(
+        &self,
+        settlement: &domain::settlement::Settlement,
+    ) -> Result<domain::settlement::Auction, error::Auction> {
+        let mut ex = self.postgres.pool.begin().await.context("begin")?;
+
+        let auction = settlement.auction_id();
+
+        let (winner, score, deadline) = {
+            let scores = database::settlement_scores::fetch(&mut ex, auction)
+            .await
+            .context("fetch scores")?
+            // if score is missing, no competition / auction exist for this auction_id
+            .ok_or(error::Auction::MissingScore)?;
+
+            (
+                H160(scores.winner.0).into(),
+                competition::Score::new(
+                    big_decimal_to_u256(&scores.winning_score)
+                        .ok_or(error::Auction::DbConversion("score"))?
+                        .into(),
+                )?,
+                (scores.block_deadline as u64).into(),
+            )
+        };
+
+        let prices = {
+            let db_prices = database::auction_prices::fetch(&mut ex, auction)
+                .await
+                .context("fetch auction prices")?;
+
+            let mut prices = HashMap::new();
+            for price in db_prices {
+                let token = eth::H160(price.token.0).into();
+                let price = big_decimal_to_u256(&price.price)
+                    .ok_or(domain::auction::InvalidPrice)
+                    .and_then(|p| domain::auction::Price::new(p.into()))
+                    .map_err(error::Auction::Price)?;
+                prices.insert(token, price);
+            }
+            prices
+        };
+
+        let (fee_policies, missing_orders) = {
+            let mut missing_orders = HashSet::new();
+            let mut fee_policies = HashMap::new();
+            for order in settlement.order_uids().cloned() {
+                match database::orders::read_order(&mut ex, &ByteArray(order.0))
+                    .await
+                    .context("fetch order")?
+                {
+                    Some(_) => {
```

Comment:
Since we only care about `.is_some()` you could safe 1 level of indentation with `if {} else {}` here.
Or even:
```rust
if let found_order.is_none() {
    missing_orders.insert(order);
    continue;
}

// handle case where order exists without extra indentation
```